fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/mixin/registry/sync/SimpleRegistryMixin.java

```diff
@@ -382,4 +414,112 @@ public void unmap(String name) throws RemapException {
 			fabric_prevEntries = null;
 		}
 	}
+
+	@Override
+	public void addAlias(Identifier old, Identifier newId) {
+		Objects.requireNonNull(old, "alias cannot be null");
+		Objects.requireNonNull(newId, "aliased id cannot be null");
+
+		if (aliases.containsKey(old)) {
+			throw new IllegalArgumentException(
+					"Tried adding %s as an alias for %s, but it is already an alias (for %s) in registry %s".formatted(
+							old,
+							newId,
+							aliases.get(old),
+							this.key
+					)
+			);
+		}
+
+		if (this.idToEntry.containsKey(old)) {
+			throw new IllegalArgumentException(
+					"Tried adding %s as an alias, but it is already present in registry %s".formatted(
+							old,
+							this.key
+					)
+			);
+		}
+
+		if (getAliased(newId).equals(old)) {
+			// since an alias corresponds to at most one identifier, this is the only way to create a cycle
+			// that doesn't already fall under the first condition
+			throw new IllegalArgumentException(
+					"Making %1$s an alias of %2$s would create a cycle, as %2$s is already an alias of %1$s (registry %3$s)".formatted(
+							old,
+							newId,
+							this.key
+					)
+			);
+		}
+
+		if (!this.idToEntry.containsKey(newId)) {
+			FABRIC_LOGGER.warn(
+					"Adding {} as an alias for {}, but the latter doesn't exist in registry {}",
+					old,
+					newId,
+					this.key
+			);
+		}
+
+		assertNotFrozen();
+		aliases.put(old, newId);
+		FABRIC_LOGGER.debug("Adding alias {} for {} in registry {}", old, newId, this.key);
+	}
+
+	@Unique
+	private Identifier getAliased(Identifier id) {
+		Identifier alias = id;
+
+		while (aliases.containsKey(alias)) {
+			alias = aliases.get(alias);
```

Comment:
You can use the fact that `get` returns `null` if the key is absent to avoid doing two map lookups. However I am thinking for performance reasons, the map should point directly to the final ID so you don't have to walk it. The hard work to "flatten" the map could be done in `addAlias` which hopefully doesn't get called as often.